[KOGITO-2851] release notes for engine modularization (#2875)
kiegroup/kie-docs: doc-content/drools-docs/src/main/asciidoc/ReleaseNotes/ReleaseNotesDrools.7.45.0.Final/drools-modularization.adoc

[id='drools-modularization']

> = Modularization of drools core engine

Reviewer: @hmanwani-rh Should be "Modularization of {DECISION_ENGINE}" ideally, but if not, at least need cap Drools here.

> = Modularization of drools core engine

Reviewer 2: @sterobin thanks, I have updated the suggestion

In order to make the core engine of drools lighter and more maintainable there is an ongoing refactoring process to isolate
and descope optional features that don't belong to internal algorithm of the engine itself. This process started with version
7.40 with the descoping of traits and continued in this release with a similar approach to isolate in separated modules both the

> features of the engine requiring mvel/asm and the ones involving protobuf based serialization.

Reviewer 2: ```suggestion
To make the core engine of {PRODUCT} lighter and maintainable, you can use the refactoring process of isolating and descoping optional features that are not related to internal algorithm of the engine. This process starts with version 7.40 with the descoping of traits and continues to the current release with a similar approach of isolating the features in separate modules. The isolated features require MVEL or ASM and involve protobuf (protocol buffer) based on serialization.
```

Author: @hmanwani-rh It is not correct to say that "the user can use this process". In other words this is not something optional, but just something to which all users should adjust their projects. We, the drools team, implemented this refactor and drools' users have to change their code to make it compatible with this change.

Reviewer 2: @mariofusco Thanks for making it clear. I have updated the suggestion, where we are describing that there is an existing refactoring process.

== Descoping of MVEL/ASM based features

Historically Drools leverages MVEL for a lot of internal usages like drl parsing, constraints evaluation, and templates generation.
Similarly ASM has been used for bytecode generation of declared types, traits and compiled contraints. The introduction of the
executable model provided an alternative (and under many points of view better) way to cover all these needs and then these 2

> tools are no longer strictly required.

Reviewer 2: ```suggestion
{PRODUCT} leverages MVEL for internal usages such as DRL parsing, constraints evaluation, and templates generation. Similarly, ASM is used for the bytecode generation of declared types, traits, and compiled constraints. The executable model provides an alternative way to cover all the requirements and then MVEL or ASM is no longer required.
```

For this reason all Drools usages of MVEL and ASM have been isolated in a new `drools-mvel` module. Having this new module on the
classpath is strictly required for all usages not involving the executable model. Note that this is unrelated with the dialect in use:
going through "traditional" drl compilation, both java and mvel dialects heavily relies on MVEL for parsing and on ASM for bytecode

> generation and in both cases it is needed to have `drools-mvel` in your classpath.

Reviewer 2: ```suggestion
Therefore, {PRODUCT} usages of MVEL and ASM are isolated in a new `drools-mvel` module. This new module on the classpath is strictly required for all usages that do not involve the executable model. Note that this is not related to the dialect in use. In the usual DRL compilation, both Java and MVEL dialects rely on MVEL for parsing and on ASM for bytecode generation. In both cases, it is a must to have `drools-mvel` in your classpath.
```

> == Descoping of protobuf based serialization features

Reviewer 2: ```suggestion
=== Descope protobuf based serialization features
```

Drools implements `KieSessions` serialization with protobuf in order to guarantee a better backward compatibility than the one
provided by the native Java serialization mechanism. However in the vast majority of use cases `KieSessions` serialization is
not required and it was not correct that this features was mixed with the core internal algorithms of the engine. This is why
this serialization feature is now isolated in the new `drools-serialization-protobuf` module. This name has been chosed also
to reflect the fact that in future Drools could provide different implementations for `KieSessions` serialization, eventually

> based on different technologies.

Reviewer 2: ```suggestion
{PRODUCT} implements `KieSessions` serialization with protobuf to ensure better backward compatibility compared to the native Java serialization mechanism. However, in the majority of use cases `KieSessions` serialization is not required and it is not correct that the serialization feature is mixed with the core internal algorithms of the engine. Therefore, the serialization feature is isolated in the new `drools-serialization-protobuf` module. `drools-serialization-protobuf` reflects the fact that {PRODUCT} can provide a different implementation for `KieSessions` serialization that is based on different technologies. 
```

> == New wrapper modules

Reviewer 2: ```suggestion
=== New wrapper modules
```

The introduction of these new modules will very likely break the biggest part of existing projects using Drools. The trivial fix
for these situations will be adding to the project's classpath `drools-mvel` and/or `drools-serialization-protobuf` according to

> the project's needs.

Reviewer 2: ```suggestion
The introduction of these new modules breaks the biggest part of existing projects using {PRODUCT}. The trivial fix for these situations is adding `drools-mvel` and `drools-serialization-protobuf` to the classpath of the project based on the project requirements.
```

This issue is mainly caused by the fact that users have among their dependencies modules like `drools-core` and `drools-compiler`
that in reality are no more than "internal implementation details". To avoid similar problems in future, in the very likely case
that other modules will be introduced to split engine in finer grained submodules and descope optional features, the following 2

> new wrapper modules have been created to cover the 2 different main usage scenarios:

Reviewer 2: ```suggestion
The reason for this issue is that the users contain dependencies modules such as `drools-core` and `drools-compiler`, which are internal implementation details. To avoid this issue, you can introduce other modules to spilt the engine in submodules and descope the optional features. Following are two new wrapper modules that are created to cover different usage scenarios:
```

. *drools-engine* aggregating `drools-core`, `drools-compiler` and `drools-model-compiler`

> . *drools-engine-classic* aggregating `drools-core`, `drools-compiler` and `drools-mvel`

Reviewer 2: ```suggestion
* *drools-engine* aggregating `drools-core`, `drools-compiler`, and `drools-model-compiler`
* *drools-engine-classic* aggregating `drools-core`, `drools-compiler`, and `drools-mvel`
```

With the introduction of these 2 modules, to use Drools with the executable model it will be enough to add `drools-engine` to a
project's dependencies while for traditional use of Drools without the model it will be necessary to just import `drools-engine-classic`.
These names reflect the intention of the Drools team to promote the use of the executable model and make it the default choice

> at least for new users.

Reviewer 2: ```suggestion
If you are using {PRODUCT} with the executable model (recommended), then add `drools-engine` module to your project dependencies. However, if you are using {PRODUCT} without the executable model, then you must import `drools-engine-classic` module. 
```
